BUG: Index.take may add unnecessary freq attribute

sinhrks opened the pull request:
`Index.freq` should raise `AttributeError` in non-datetime-like index. But `Index.take` adds `freq=None` and broke the behavior. 

```
import pandas as pd
idx = pd.Index([1, 2, 3])

idx.freq
# AttributeError: 'Int64Index' object has no attribute 'freq'

idx.take([True, False, True]).freq
# None
```

sinhrks committed:
BUG: Index.take may add unnecessary freq attribute

jreback commented on pandas/core/index.py:
```diff
@@ -1203,11 +1203,9 @@ def take(self, indexer, axis=0):
-        # by definition cannot propogate freq
-        return self._shallow_copy(taken, freq=None)
+        indices = com._ensure_platform_int(indices)
+        taken = np.array(self).take(indices)
```
I realize that this is actually a perf issue coming from here
do the take directly in self.values as take always copies anyhow
(this may fail for categoricla) and u might be able to get away with np.array(arr, copy=False)

sinhrks commented:
@jreback Thanks for the great advice. Fixed to use ``.values.take`` and ``.asi8.take`` in datetime-like case.

Mainly ``iloc`` gets faster.  I've also added benches for ``take`` (``series_take_intindex`` and ``series_take_dtindex``), but results look not significant in these cases. ``series_xs_mi_ix`` and ``frame_xs_mi_ix`` results look random.

If it looks reasonable, I'll add release not for perf improvement also.

```
-------------------------------------------------------------------------------
Test name                                    | head[ms] | base[ms] |  ratio   |
-------------------------------------------------------------------------------
series_iloc_list_like                        |   0.1010 |   0.8353 |   0.1209 |
series_xs_mi_ix                              |   1.2943 |   2.4173 |   0.5354 |
frame_xs_mi_ix                               |   1.3490 |   2.4823 |   0.5434 |
series_iloc_array                            |   3.4820 |   4.0780 |   0.8539 |
frame_reindex_upcast                         |   5.9937 |   6.9266 |   0.8653 |
frame_ctor_list_of_dict                      |  59.9620 |  67.9890 |   0.8819 |
frame_dropna_axis1_all                       |  68.5070 |  75.5123 |   0.9072 |
frame_dropna_axis0_any                       |  22.2783 |  23.9987 |   0.9283 |
series_take_intindex                         |   2.7970 |   3.0077 |   0.9300 |
strings_startswith                           |   3.6437 |   3.8893 |   0.9368 |
strings_endswith                             |   3.6343 |   3.8550 |   0.9428 |
series_take_dtindex                          |   2.7507 |   2.9174 |   0.9429 |
...
timeseries_timestamp_tzinfo_cons             |   0.0106 |   0.0090 |   1.1858 |
timeseries_infer_freq                        |  11.5213 |   9.6280 |   1.1966 |
timeseries_is_month_start                    |   3.1753 |   2.6217 |   1.2112 |
timeseries_custom_bday_cal_incr_n            |   0.0410 |   0.0330 |   1.2434 |
-------------------------------------------------------------------------------
Test name                                    | head[ms] | base[ms] |  ratio   |
-------------------------------------------------------------------------------
```

jreback commented:
@sinhrks sounds good.

sinhrks commented:
Updated the release note. Pls review.

jreback commented:
lgtm
